mitiq/vd/vd_utils.py

```diff
@@ -51,3 +55,137 @@ def map_for_grid_qubits(qu: cirq.Qid) -> cirq.Qid:
+        matrix[:] = matrix[:, idx]
+    else:
+        raise TypeError("matrix should be 2D or 3D ndarray")
+    return matrix
+
+
+def _apply_symmetric_observable(
+    matrix: NDArray[np.complex128],
+    N_qubits: int,
+    observable: Optional[Union[Observable, NDArray[np.complex128]]] = None,
+    M: int = 2,
+) -> NDArray[np.complex128]:
+    """
+    Function that applies a symmetric observable to a matrix or vector.
+    The observable is assumed to be the Z observable by default.
+
+    Args:
+        matrix: The matrix or vector that should be shifted.
+        N_qubits: The number of qubits in each register.
+        observable: The observable that should be applied.
+        If None, the Z observable is used.
+        M: The number of registers.
+
+    Returns:
+        The matrix or vector with the observable applied.
+    """
+    z_matrix = np.array([[1.0, 0.0], [0.0, -1.0]])
+
+    if observable is None or (
+        isinstance(observable, np.ndarray)
+        and np.allclose(observable, z_matrix)
+    ):
+        # use the default Z observable
+        sym_observable_diagonals: List[NDArray[np.complex128]] = []
+        for i in range(N_qubits):
+            observable_i_diagonal = np.array(
+                [
+                    j
+                    for k in range(2 ** (i))
+                    for j in [1.0, -1.0]
+                    for ll in range(2 ** (N_qubits - i - 1))
+                ]
+            )
+
+            # turn [a, b, c] into [a,a,a,b,b,b,c,c,c]. This is the same as
+            # tensoring the N_qubit identity on the right
+            observable_i_diagonal_system1 = np.array(
+                [observable_i_diagonal for _ in range(2**N_qubits)]
+            ).flatten("F")
+            # turn [a,b,c] into [a,b,c,a,b,c,a,b,c]. This is the same as
+            # tensoring the N_qubit identity on the left
+            observable_i_diagonal_system2 = np.array(
+                [observable_i_diagonal for _ in range(2**N_qubits)]
+            ).flatten("C")
+            # add the symmetric observable
+            # Create the combined diagonal and explicitly convert to complex128
+            combined_diagonal = 0.5 * (
+                observable_i_diagonal_system1 + observable_i_diagonal_system2
+            )
+            sym_observable_diagonals.append(
+                combined_diagonal.astype(np.complex128)
+            )
+
+        if matrix.ndim == 2:
+            return np.array([sod * matrix for sod in sym_observable_diagonals])
+        elif matrix.ndim == 3:
+            return np.array(
+                [
+                    sod * mat
+                    for sod in sym_observable_diagonals
+                    for mat in matrix
+                ]
+            )
+        else:
+            raise ValueError("matrix should be 2D or 3D ndarray")
+
+    else:
+        obs_array = (
+            observable
+            if isinstance(observable, np.ndarray)
+            else np.array(observable)
```

Comment:
```suggestion
            else observable.matrix()
```

`np.array(observable)` creates some array based on ... something?

```py
>>> ob = Observable(PauliString("XY"))
>>> np.array(ob)
array(<mitiq.observable.observable.Observable object at 0x16a872930>,
      dtype=object)
```

Reply:
good question, I am assuming they expect a matrix in list form? which doesn't make sense because the type hint suggests either np array or Observable. so instead of np.array(ob) we should use the matrix form of the Observable object. what do you think?

Reply:
Yup that's what I think as well! Thanks for making the change.